CRM/Core/EntityTokens.php

```diff
@@ -25,31 +32,29 @@
  * AbstractTokenSubscriber in future. It is being used to clarify
  * functionality but should NOT be used from outside of core tested code.
  */
-class CRM_Core_EntityTokens extends AbstractTokenSubscriber {
+abstract class CRM_Core_EntityTokens extends AbstractTokenSubscriber {
 
   /**
    * @inheritDoc
    * @throws \CRM_Core_Exception
    */
   public function evaluateToken(TokenRow $row, $entity, $field, $prefetch = NULL) {
-    $fieldValue = $this->getFieldValue($row, $field);
-
-    if ($this->isPseudoField($field)) {
-      $split = explode(':', $field);
-      return $row->tokens($entity, $field, $this->getPseudoValue($split[0], $split[1], $this->getFieldValue($row, $split[0])));
+    $aliases = $this->getAliasTokens();
+    if (isset($aliases[$field])) {
+      $this->evaluateToken($row, $entity, $aliases[$field], $prefetch);
+      return $row->copyToken("$entity." . $aliases[$field], "$entity.$field");
     }
-    if ($this->isMoneyField($field)) {
```

Comment:
This is generic - lots of classes will need this - so moving it to contribution is the wrong way - we might need a 'getCurrency' function for when we hit a class that doesn't have it's own currency field